<a href="https://colab.research.google.com/github/juanjosedelgado-coder/Talleres/blob/main/Tercer_Parcial_Seleccion_Mejor_Modelo_Vargas_Cuaspud_Delgado.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<a href="https://colab.research.google.com/github/LinaMariaCastro/curso-ia-para-economia/blob/main/clases/5_Aprendizaje_supervisado/6_Competencia_Seleccion_Mejor_Modelo.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Inteligencia Artificial con Aplicaciones en Economía I**

- 👩‍🏫 **Profesora:** [Lina María Castro](https://www.linkedin.com/in/lina-maria-castro)  
- 📧 **Email:** [lmcastroco@gmail.com](mailto:lmcastroco@gmail.com)  
- 🎓 **Universidad:** Universidad Externado de Colombia - Facultad de Economía

# 🏆 **Tercer parcial: Selección del Mejor Modelo**

**ESTÁ PROHIBIDO EL USO DE GRANDES MODELOS DE LENGUAJE COMO CHATGPT, CLAUDE, GEMINI, ENTRE OTROS, PARA RESOLVER ESTE EJERCICIO**

**Trabajo en grupos de 3**

**Objetivo:** Predecir las ventas de una compañía (`Sales`) teniendo en cuenta su inversión en publicidad.

**Dataset:** `train_df_ventas.csv` disponible en el repositorio del curso.

**IMPORTANTE: Los datos cargados solo corresponden a `train`.**

**Metodología:**
1.  Cargar y explorar los datos.
2.  Preprocesar los datos si es necesario.
3.  De los siguientes modelos, entrenar por lo menos 2:
    - Regresión Lineal
    - Regresión Polinómica
    - KNN Regressor
    - Decision Tree Regressor
    - Random Forest Regressor
    - Gradient Boosting Regressor
    - XGBoost Regressor
4.  Si lo considera necesario, usar `GridSearchCV` con Validación Cruzada (`cv=5`) para optimizar los hiperparámetros. La métrica de optimización debe ser el **RMSE** (Root Mean Squared Error), por lo que debe usar `scoring='neg_root_mean_squared_error'` (el valor será negativo y se multiplicará por -1 al final).
5.  Comparar los modelos y seleccionar el mejor, teniendo en cuenta el menor RMSE.

**Forma de entrega:**

- Nombrar el archivo de la siguiente forma: “Tercer_Parcial_apellidos.ipynb”.
- Suba el Jupyter Notebook a su cuenta en Github y envíe el link en el siguiente Forms: https://forms.cloud.microsoft/r/Bsy2U83tbc. No olvide indicar claramente cuál es el modelo seleccionado.

**IMPORTANTE:** No se recibirán talleres en Google Colab, el notebook debe estar subido en Github.

**Calificación**

La docente, evaluará el modelo seleccionado por ustedes en el `test set`.

El proceso seguirá estas reglas:

- **Criterio de Ganador:** El equipo que tenga todo el procedimiento correcto y obtenga el Root Mean Squared Error (RMSE) más bajo en el test set recibirá una calificación de 5.0.

- **Criterio de Desempate:** En caso de empate en el RMSE, se otorgará la ventaja al equipo que haya entrenado y evaluado más modelos.

- **Escalafón de Notas:** A partir del primer puesto, se restará 0.1 a la nota final por cada posición inferior (2º lugar: 4.9, 3er lugar: 4.8, etc.).

- **Validación de Procedimiento:** Es obligatorio que el código sea reproducible por la docente (no olivde colocar las semillas en los procesos aleatorios). Si el script contiene errores, el equipo quedará fuera de la competencia y se dará una calificación acorde a lo que esté correcto.

**Explicación de las variables:**

- Sales: Ventas (millones USD). --> **Esta es la variable objetivo**
- TV: Gasto en promoción televisiva (millones USD).
- Radio: Gasto en promoción radiofónica (millones USD).
- Social Media: Gasto en promoción en redes sociales (millones USD).
- Influencer: Indica si la promoción se realizó en colaboración con Mega, Macro, Nano o Micro influencers.


**Nombres estudiantes del equipo:**

- Natalia vargas
- Marleny Cuaspud
- Juan Jose Delgado

# **Desarrollo**

In [1]:
import os
import warnings
warnings.filterwarnings('ignore')

# Manipulación de datos
import pandas as pd
import numpy as np

# Visualización
import matplotlib.pyplot as plt
import seaborn as sns

# Preprocesamiento
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer

# Modelos
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix, roc_curve, auc, precision_recall_curve
import shap
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.metrics import mean_squared_error, r2_score

# Métricas y Tuning
from sklearn.model_selection import GridSearchCV, cross_val_score
from sklearn.metrics import classification_report, confusion_matrix, roc_auc_score, roc_curve

In [2]:
# Que muestre todas las columnas
pd.options.display.max_columns = None
# En los dataframes, mostrar los float con dos decimales
pd.options.display.float_format = '{:,.2f}'.format

# Configuraciones para una mejor visualización
sns.set(style='whitegrid')
plt.rcParams['figure.figsize'] = (10, 6)

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [4]:
df = pd.read_csv('/content/drive/MyDrive/IA Economia/train_df_ventas.csv')

df.head()

,TV,Social Media,Influencer,Radio,Sales
0,17.58,1.22,Macro,22.28,173.61
1,14.59,5.48,Micro,22.10,140.54
2,25.18,2.28,Mega,22.20,197.35
3,12.90,1.83,Nano,21.74,184.05
4,28.38,4.34,Nano,22.61,318.63


In [5]:
df.info()
df.describe()
df.isnull().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3636 entries, 0 to 3635
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   TV            3636 non-null   float64
 1   Social Media  3636 non-null   float64
 2   Influencer    3636 non-null   object 
 3   Radio         3636 non-null   float64
 4   Sales         3636 non-null   float64
dtypes: float64(4), object(1)
memory usage: 142.2+ KB


,0
TV,0
Social Media,0
Influencer,0
Radio,0
Sales,0


In [6]:
x = df.drop('Sales', axis=1)
y = df['Sales']

In [7]:
x_train, x_test, y_train, y_test = train_test_split(
    x,
    y,
    test_size=0.2,
    random_state=42
)

In [8]:
numerical_features = x.select_dtypes(include=["int64","float64"]).columns
categorical_features = x.select_dtypes(include="object").columns

print(f"Columnas Numéricas: {list(numerical_features)}")
print(f"Columnas Categóricas: {list(categorical_features)}")

Columnas Numéricas: ['TV', 'Social Media', 'Radio']
Columnas Categóricas: ['Influencer']


In [9]:
# Crear el transformador para las categóricas
categorical_transformer = OneHotEncoder(handle_unknown='ignore')
# Crear el preprocesador (ColumnTransformer)
preprocessor = ColumnTransformer(
    transformers=[
        ('cat', categorical_transformer, categorical_features)
    ], remainder= 'passthrough')

In [10]:
# Ajustamos (fit) y transformamos el set de entrenamiento

x_train_processed = preprocessor.fit_transform(x_train)

# SÓLO transformamos el set de prueba (ya "aprendió" del train)

x_test_processed = preprocessor.transform(x_test)

print(f"\nForma de X_train procesado: {x_train_processed.shape}")
print(f"Forma de X_test procesado: {x_test_processed.shape}")


Forma de X_train procesado: (2908, 7)
Forma de X_test procesado: (728, 7)


In [11]:
# Inicializa el modelo, usa DecisionTreeRegressor
# Coloca como semilla random_state=42

tree_model = DecisionTreeRegressor(random_state=42)

# Entrena el modelo

tree_model.fit(x_train_processed, y_train)

# Realiza predicciones en el conjunto de prueba

y_pred_tree = tree_model.predict(x_test_processed)

In [12]:
# Definimos la grilla de hiperparámetros
param_grid_tree = {
    'max_depth': [3, 5, 10, None],
    'min_samples_leaf': [1, 2, 5, 10],
    'min_samples_split': [2, 5, 10]
}

# Creamos el modelo base
tree_model_grid = DecisionTreeRegressor(random_state=42)

# Configuramos GridSearchCV
grid_tree = GridSearchCV(
    estimator=tree_model_grid,
    param_grid=param_grid_tree,
    cv=5,
    scoring='neg_root_mean_squared_error',
    n_jobs=-1
)

# Entrenamos
grid_tree.fit(x_train_processed, y_train)

GridSearchCV(cv=5, estimator=DecisionTreeRegressor(random_state=42), n_jobs=-1,
             param_grid={'max_depth': [3, 5, 10, None],
                         'min_samples_leaf': [1, 2, 5, 10],
                         'min_samples_split': [2, 5, 10]},
             scoring='neg_root_mean_squared_error')

In [13]:
print("Mejores hiperparámetros:")
print(grid_tree.best_params_)

Mejores hiperparámetros:
{'max_depth': 5, 'min_samples_leaf': 1, 'min_samples_split': 2}


In [14]:
print("Mejor RMSE:")
print(-grid_tree.best_score_)

Mejor RMSE:
44.25096796836626


In [15]:
-grid_tree.best_score_

np.float64(44.25096796836626)

In [16]:
best_tree_model = grid_tree.best_estimator_

In [17]:
y_pred_best_tree = best_tree_model.predict(x_test_processed)

In [18]:
rmse_tree = np.sqrt(mean_squared_error(y_test, y_pred_best_tree))

print("RMSE final:", rmse_tree)

RMSE final: 44.443560557735886


In [19]:
# Modelo Random Forest

rf_model = RandomForestRegressor(random_state=42)

# Hiperparámetros
param_grid_rf = {
    'n_estimators': [100, 200],
    'max_depth': [5, 10, None],
    'min_samples_leaf': [1, 2, 5]
}

# GridSearchCV
grid_rf = GridSearchCV(
    estimator=rf_model,
    param_grid=param_grid_rf,
    cv=5,
    scoring='neg_root_mean_squared_error',
    n_jobs=-1
)

In [20]:
# Entrenar
grid_rf.fit(x_train_processed, y_train)

GridSearchCV(cv=5, estimator=RandomForestRegressor(random_state=42), n_jobs=-1,
             param_grid={'max_depth': [5, 10, None],
                         'min_samples_leaf': [1, 2, 5],
                         'n_estimators': [100, 200]},
             scoring='neg_root_mean_squared_error')

In [21]:
# Mejor modelo
best_rf = grid_rf.best_estimator_

# Predicciones
y_pred_rf = best_rf.predict(x_test_processed)

In [22]:
rmse_rf = np.sqrt(mean_squared_error(y_test, y_pred_rf))

print("RMSE Random Forest:", rmse_rf)

RMSE Random Forest: 42.68036815081386


In [23]:
print("RMSE Árbol:", rmse_tree)
print("RMSE Random Forest:", rmse_rf)

RMSE Árbol: 44.443560557735886
RMSE Random Forest: 42.68036815081386


In [25]:
scoring='neg_root_mean_squared_error'

##Gradient boosting

In [26]:
gb_model = GradientBoostingRegressor(random_state=42)

In [27]:
param_grid_gb = {
    'n_estimators': [100, 200],
    'learning_rate': [0.05, 0.1],
    'max_depth': [3, 5]
}

In [28]:
grid_gb = GridSearchCV(
    estimator=gb_model,
    param_grid=param_grid_gb,
    cv=3,
    scoring='neg_root_mean_squared_error',
    n_jobs=-1
)

In [29]:
grid_gb.fit(x_train_processed, y_train)

GridSearchCV(cv=3, estimator=GradientBoostingRegressor(random_state=42),
             n_jobs=-1,
             param_grid={'learning_rate': [0.05, 0.1], 'max_depth': [3, 5],
                         'n_estimators': [100, 200]},
             scoring='neg_root_mean_squared_error')

In [30]:
print("Mejores hiperparámetros:")
print(grid_gb.best_params_)

Mejores hiperparámetros:
{'learning_rate': 0.05, 'max_depth': 3, 'n_estimators': 100}


In [31]:
print("Mejor RMSE CV:")
print(-grid_gb.best_score_)

Mejor RMSE CV:
42.42610116244159


In [32]:
best_gb = grid_gb.best_estimator_

In [33]:
y_pred_gb = best_gb.predict(x_test_processed)

In [34]:
rmse_gb = np.sqrt(mean_squared_error(y_test, y_pred_gb))

print("RMSE Gradient Boosting:", rmse_gb)

RMSE Gradient Boosting: 42.397527881057265


In [35]:
r2_gb = r2_score(y_test, y_pred_gb)

print("R² Gradient Boosting:", r2_gb)

R² Gradient Boosting: 0.7891945320942251


In [36]:
print("RMSE Árbol:", rmse_tree)
print("RMSE Random Forest:", rmse_rf)
print("RMSE Gradient Boosting:", rmse_gb)

RMSE Árbol: 44.443560557735886
RMSE Random Forest: 42.68036815081386
RMSE Gradient Boosting: 42.397527881057265


In [38]:
results = pd.DataFrame({
    'Modelo': ['Árbol de Decisión', 'Random Forest', 'Gradient Boosting'],

    'RMSE': [
        np.sqrt(mean_squared_error(y_test, y_pred_best_tree)),
        np.sqrt(mean_squared_error(y_test, y_pred_rf)),
        np.sqrt(mean_squared_error(y_test, y_pred_gb))
    ],

    'R2': [
        r2_score(y_test, y_pred_best_tree),
        r2_score(y_test, y_pred_rf),
        r2_score(y_test, y_pred_gb)
    ]
})

results = results.sort_values(by='RMSE')

results

,Modelo,RMSE,R2
2,Gradient Boosting,42.40,0.79
1,Random Forest,42.68,0.79
0,Árbol de Decisión,44.44,0.77


# **Indica claramente cuál modelo seleccionaste como el mejor**

**R/**

El modelo seleccionado como el mejor fue Gradient Boosting Regressor, ya que obtuvo el menor RMSE en el conjunto de prueba (42.40), superando al Random Forest Regressor (42.68) y al Decision Tree Regressor (44.44). Esto indica que Gradient Boosting fue el modelo que logró predecir las ventas con menor error promedio. Su mejor desempeño se debe a que construye árboles de manera secuencial, donde cada nuevo árbol corrige los errores cometidos por los anteriores, permitiendo capturar patrones más complejos en los datos y mejorar la capacidad de generalización del modelo. Por esta razón, se considera el modelo más adecuado para predecir las ventas de la compañía a partir de la inversión en publicidad.